rust/connlib/clients/shared/src/eventloop.rs

```diff
@@ -0,0 +1,467 @@
+use crate::{
+    messages::{
+        BroadcastGatewayIceCandidates, Connect, ConnectionDetails, EgressMessages,
+        GatewayIceCandidates, IngressMessages, InitClient, RemoveResource, ReplyMessages,
+    },
+    PHOENIX_TOPIC,
+};
+use anyhow::{anyhow, Context as _, Result};
+use async_compression::tokio::bufread::GzipEncoder;
+use bimap::BiMap;
+use connlib_shared::{
+    messages::{ConnectionAccepted, DnsServer, GatewayResponse, IpDnsServer, ResourceAccepted},
+    Callbacks, IpProvider,
+};
+use firezone_tunnel::ClientTunnel;
+use ip_network::IpNetwork;
+use phoenix_channel::PhoenixChannel;
+use reqwest::header::{CONTENT_ENCODING, CONTENT_TYPE};
+use std::{
+    convert::Infallible,
+    io,
+    net::IpAddr,
+    path::PathBuf,
+    str::FromStr,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::{
+    io::BufReader,
+    time::{Instant, Interval, MissedTickBehavior},
+};
+use tokio_util::codec::{BytesCodec, FramedRead};
+use url::Url;
+
+const DNS_PORT: u16 = 53;
+const DNS_SENTINELS_V4: &str = "100.100.111.0/24";
+const DNS_SENTINELS_V6: &str = "fd00:2021:1111:8000:100:100:111:0/120";
+
+pub struct Eventloop<C: Callbacks> {
+    tunnel: ClientTunnel<C>,
+    tunnel_init: bool,
+
+    portal: PhoenixChannel<(), IngressMessages, ReplyMessages>,
+
+    print_stats_timer: tokio::time::Interval,
+}
+
+impl<C: Callbacks> Eventloop<C> {
+    pub(crate) fn new(
+        tunnel: ClientTunnel<C>,
+        portal: PhoenixChannel<(), IngressMessages, ReplyMessages>,
+    ) -> Self {
+        Self {
+            tunnel,
+            portal,
+            print_stats_timer: tokio::time::interval(Duration::from_secs(10)),
+            tunnel_init: false,
+        }
+    }
+}
+
+impl<C> Eventloop<C>
+where
+    C: Callbacks + 'static,
+{
+    #[tracing::instrument(name = "Eventloop::poll", skip_all, level = "debug")]
+    pub fn poll(&mut self, cx: &mut Context<'_>) -> Poll<Result<Infallible>> {
+        loop {
+            match self.tunnel.poll_next_event(cx)? {
+                Poll::Ready(firezone_tunnel::Event::SignalIceCandidate {
+                    conn_id: gateway,
+                    candidate,
+                }) => {
+                    tracing::debug!(%gateway, %candidate, "Sending ICE candidate to client");
+
+                    self.portal.send(
+                        PHOENIX_TOPIC,
+                        EgressMessages::BroadcastIceCandidates(BroadcastGatewayIceCandidates {
+                            gateway_ids: vec![gateway],
+                            candidates: vec![candidate],
+                        }),
+                    );
+
+                    continue;
+                }
+                Poll::Ready(firezone_tunnel::Event::ConnectionIntent {
+                    connected_gateway_ids,
+                    resource,
+                    ..
```

Comment:
@conectado Behind this `..` is hiding a `reference` parameter. What is the purpose of passing out a reference here? `phoenix-channel` does not support setting the reference because it generates an `OutboundRequestId` for you that uniquely identifies the request.

Reply:
Summary from a discussion on how to handle this:

- Replace the interval with a debounce, only generate a new connection intent every 2 seconds in case the client retries.
- Don't pass a reference out with the `ConnectionIntent`, instead, generate an ID before sending it via the phoenix-channel and remember which resource ID this was for.
- When receiving the response, check if the response is the highest (i.e. last) intent we sent and only pass it on then.

Do all of the above in a separate PR on main to make integration here easier.